core/services/ocrcommon/adapters.go

```diff
@@ -71,3 +76,66 @@ func (c *OCR3ContractTransmitterAdapter) Transmit(ctx context.Context, digest oc
 func (c *OCR3ContractTransmitterAdapter) FromAccount() (ocrtypes.Account, error) {
 	return c.ct.FromAccount()
 }
+
+var _ ocr3types.OnchainKeyring[[]byte] = (*OCR3OnchainKeyringMultiChainAdapter)(nil)
+
+type OCR3OnchainKeyringMultiChainAdapter struct {
+	ks         keystore.OCR2
+	st         ocr2.OCR2OnchainSigningStrategy
+	keyBundles map[string]ocr2key.KeyBundle
+	publicKey  ocrtypes.OnchainPublicKey
+}
+
+func NewOCR3OnchainKeyringMultiChainAdapter(ks keystore.OCR2, st ocr2.OCR2OnchainSigningStrategy) (*OCR3OnchainKeyringMultiChainAdapter, error) {
+	var keyBundles map[string]ocr2key.KeyBundle
+	for chainFamily, _ := range types.SupportedRelays {
+		kbID, err := st.KeyBundleID(chainFamily)
+		if err != nil {
+			return nil, err
+		}
+		kb, err := ks.Get(kbID)
+		if err != nil {
+			return nil, err
+		}
+		keyBundles[chainFamily] = kb
+	}
+	pkKeyBundleID, _ := st.PublicKey()
+	kb, err := ks.Get(pkKeyBundleID)
+	if err != nil {
+		return nil, err
+	}
+	return &OCR3OnchainKeyringMultiChainAdapter{ks, st, keyBundles, kb.PublicKey()}, nil
+}
+
+func (a *OCR3OnchainKeyringMultiChainAdapter) PublicKey() ocrtypes.OnchainPublicKey {
+	return a.publicKey
+}
+
+func (a *OCR3OnchainKeyringMultiChainAdapter) Sign(digest ocrtypes.ConfigDigest, seqNr uint64, r ocr3types.ReportWithInfo[[]byte]) (signature []byte, err error) {
+	kb := a.keyBundles[types.NetworkEVM] // TODO: how do we get the bundle name from the report info?
+	return kb.Sign(ocrtypes.ReportContext{
+		ReportTimestamp: ocrtypes.ReportTimestamp{
+			ConfigDigest: digest,
+			Epoch:        uint32(seqNr),
+			Round:        0,
+		},
+		ExtraHash: [32]byte(make([]byte, 32)),
+	}, r.Report)
+}
+
+func (a *OCR3OnchainKeyringMultiChainAdapter) Verify(opk ocrtypes.OnchainPublicKey, digest ocrtypes.ConfigDigest, seqNr uint64, ri ocr3types.ReportWithInfo[[]byte], signature []byte) bool {
```

Comment:
@vyzaldysanchez If you unmarshal ri into a structpb.Struct and then extract a specific key from it, that should do what you want, eg. (keyBundleIdentifier = types.NetworkEVM).

```
info := structpb.Struct{}
err := proto.Unmarshal(ri.Info, &info)
if err != nil { // handle }

infoMap := info.AsMap // map[string]any
keyBundleId = infoMap["keyBundleIdentifier"]

selected, ok := a.keyBundles(keyBundleId)
if !ok { // handle incorrect identifier }

return selected.Sign...
```